src/net/send_recv/msg.rs

```diff
@@ -0,0 +1,687 @@
+    flags: SendFlags,
+) -> io::Result<usize> {
+    match addr {
+        None => backend::net::syscalls::sendmsg_noaddr(socket.as_fd(), iov, control, flags),
+        Some(SocketAddrAny::V4(addr)) => {
+            backend::net::syscalls::sendmsg_v4(socket.as_fd(), addr, iov, control, flags)
+        }
+        Some(SocketAddrAny::V6(addr)) => {
+            backend::net::syscalls::sendmsg_v6(socket.as_fd(), addr, iov, control, flags)
+        }
+        #[cfg(unix)]
+        Some(SocketAddrAny::Unix(addr)) => {
+            backend::net::syscalls::sendmsg_unix(socket.as_fd(), addr, iov, control, flags)
+        }
+    }
+}
+
+/// `recvmsg(msghdr)`- Receives a message from a socket.
+pub fn recvmsg(
+    socket: impl AsFd,
+    iov: &mut [io::IoSliceMut<'_>],
+    control: &mut RecvAncillaryBuffer<'_>,
+    flags: RecvFlags,
+) -> io::Result<RecvMsgResult> {
+    backend::net::syscalls::recvmsg(socket.as_fd(), iov, control, flags)
+}
+
+/// The result of a `recvmsg` call.
+#[non_exhaustive]
+pub struct RecvMsgResult {
+    /// The number of bytes received.
+    pub bytes: usize,
+
+    /// The flags received.
+    pub flags: RecvFlags,
+
+    /// The address of the socket we received from, if any.
+    pub address: Option<SocketAddrAny>,
+}
+
+/// An iterator over data in an ancillary buffer.
+pub struct AncillaryIter<'data, T> {
+    /// The data we're iterating over.
+    data: &'data mut [u8],
+
+    /// The raw data we're removing.
+    _marker: PhantomData<T>,
+}
+
+#[allow(unsafe_code)]
+impl<'data, T> AncillaryIter<'data, T> {
+    /// Create a new iterator over data in an ancillary buffer.
+    ///
+    /// # Safety
+    ///
+    /// This can only be called if we are sure that the data is contained in a
+    /// valid ancillary buffer.
+    unsafe fn new(data: &'data mut [u8]) -> Self {
+        assert_eq!(data.len() % size_of::<T>(), 0);
+
+        Self {
+            data,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Drop `n` items from this iterator.
+    unsafe fn drop_items(&mut self, n: usize) {
+        /// On drop, move the slice forward by `size_of<T>()`.
+        struct MoveForward<'a, 'b, T>(&'a mut &'b mut [u8], PhantomData<T>);
+
+        impl<T> Drop for MoveForward<'_, '_, T> {
+            fn drop(&mut self) {
+                // Move slice forward.
+                let slice = mem::take(self.0);
+                *self.0 = &mut slice[size_of::<T>()..];
+            }
+        }
+
+        if !mem::needs_drop::<T>() {
+            return;
+        }
+
+        for _ in 0..n {
+            // See if there is a `T` left.
+            if self.data.len() < size_of::<T>() {
+                return;
+            }
+
+            // Move forward by one after drop, even on panic.
+            let move_forward = MoveForward::<'_, '_, T>(&mut self.data, PhantomData);
+
+            // Drop the `T`.
+            drop(ptr::read_unaligned(move_forward.0.as_ptr() as *const T));
```

Comment:
Works by me. Originally it used `drop_in_place` to avoid a read, but eventually I realized I needed to do `read_unaligned` and had to rework it around that. Probably best to just remove it and use `self.for_each(drop)` to drop.